client/peerset/src/protocol_controller.rs

```diff
@@ -0,0 +1,559 @@
+			self.reject_connection(incoming_index);
+			return
+		}
+
+		// Check if the node is reserved first.
+		if let Some(state) = self.reserved_nodes.get_mut(&peer_id) {
+			match state {
+				// If we're already connected, don't answer, as the docs mention.
+				PeerState::Connected(_) => {},
+				PeerState::NotConnected => {
+					// It's questionable whether we should check a reputation of reserved node.
+					// FIXME: unable to call `self.is_banned()` because of borrowed `self`.
+					if self.peerset_handle.is_banned(peer_id) {
+						self.reject_connection(incoming_index);
+					} else {
+						*state = PeerState::Connected(Direction::Inbound);
+						self.accept_connection(incoming_index);
+					}
+				},
+			}
+			return
+		}
+
+		// Adding incoming peer to our set of peers even before we decide whether to accept
+		// it is questionable, but this is how it was implemented in the original `Peerset`.
+		let state = self.nodes.entry(peer_id).or_insert(PeerState::NotConnected);
+		match state {
+			// If we're already connected, don't answer, as the docs mention.
+			PeerState::Connected(_) => return,
+			PeerState::NotConnected => {
+				// FIXME: unable to call `self.is_banned()` because of borrowed `self`.
+				if self.peerset_handle.is_banned(peer_id) {
+					self.reject_connection(incoming_index);
+					return
+				}
+
+				if self.num_in >= self.max_in {
+					self.reject_connection(incoming_index);
+					return
+				}
+
+				self.num_in += 1;
+				*state = PeerState::Connected(Direction::Inbound);
+				self.accept_connection(incoming_index);
+			},
+		}
+	}
+
+	/// Returns Err(PeerId) if peer wasn't connected.
+	fn on_peer_dropped(&mut self, peer_id: PeerId, reason: DropReason) -> Result<(), PeerId> {
+		let mut is_reserved = true;
+		let state = self
+			.reserved_nodes
+			.get_mut(&peer_id)
+			.or_else(|| {
+				is_reserved = false;
+				self.nodes.get_mut(&peer_id)
+			})
+			.ok_or(peer_id)?;
+		match state {
+			PeerState::Connected(d) => {
+				if is_reserved {
+					*state = PeerState::NotConnected;
+				} else {
+					match d {
+						Direction::Inbound => self.num_in -= 1,
+						Direction::Outbound => self.num_out -= 1,
+					}
+					if let DropReason::Refused = reason {
+						self.nodes.remove(&peer_id);
+					} else {
+						*state = PeerState::NotConnected;
+					}
+				}
+				self.report_disconnect(peer_id);
+				Ok(())
+			},
+			PeerState::NotConnected => Err(peer_id),
+		}
+	}
+
+	fn alloc_slots(&mut self) {
+		if self.num_out >= self.max_out {
+			return
+		}
+
+		// Try connecting to reserved nodes first.
+		self.reserved_nodes
+			.iter_mut()
+			.filter_map(|(peer_id, state)| {
+				(matches!(state, PeerState::NotConnected) &&
+					!self.peerset_handle.is_banned(*peer_id))
+				.then_some({
+					*state = PeerState::Connected(Direction::Outbound);
+					peer_id
+				})
+			})
+			.cloned()
+			.collect::<Vec<_>>()
+			.into_iter()
+			.map(|peer_id| {
```

Comment:
I think you need to use `for_each()` because this `map()` is never called as you're never collecting the results.

Reply:
Done